collector/pg_replication_slot.go

```diff
@@ -131,6 +153,28 @@ func (PGReplicationSlotCollector) Update(ctx context.Context, instance *instance
 			pgReplicationSlotIsActiveDesc,
 			prometheus.GaugeValue, isActiveValue, slotNameLabel, slotTypeLabel,
 		)
+
+		if safeWalSize.Valid {
+			ch <- prometheus.MustNewConstMetric(
+				pgReplicationSlotSafeWal,
+				prometheus.GaugeValue, float64(safeWalSize.Int64), slotNameLabel, slotTypeLabel,
+			)
+		}
+
+		if walStatus.Valid {
+			// See https://www.postgresql.org/docs/14/view-pg-replication-slots.html
+			walStatusMap := map[string]int{
+				"reserved":   0, // reserved means that the claimed files are within max_wal_size.
+				"extended":   1, // extended means that max_wal_size is exceeded but the files are still retained, either by the replication slot or by wal_keep_size.
+				"unreserved": 2, // unreserved means that the slot no longer retains the required WAL files and some of them are to be removed at the next checkpoint. This state can return to reserved or extended.
+				"lost":       3, // lost means that some required WAL files have been removed and this slot is no longer usable.
```

Comment:
We generally try to avoid enum style metrics in Prometheus. We don't really have a first-class enum concept. (I wish we did)

There are two options I would recommend:
* Have a `wal_status` metric with a `wal_status="FOO"` metric with a value of 1.
* Have a `wal_status` metric with all valid statuses, but boolean for each state.
Or,

Reply:
yeah really sad. i was hoping to do a wal_status > 0 alert, but have to agree its hard to document theses numbers meaning. Anyway, I followed your first suggestion.

Reply:
With the state set mode, you can do an alert like `sum without (wal_status) (pg_replication_wal_status) > 0`. So that should work exactly as you want.

Reply:
With the current method, your alert would be also simple:

```
pg_replication_wal_status{wal_status!="reserved"} == 1
```

This has the advantage of also providing the status as a string to your alert labels template.